File: 05_Development/CareBridgeAITriageService/app/context/complaint_taxonomy.py
"""Deterministic out-of-scope complaint classification.

OUT_OF_SCOPE requires **positive evidence** that a complaint belongs to another clinical
domain. An unrecognised complaint is NEEDS_MORE_INFO — "we did not understand you" is not the
same statement as "this is not our field", and conflating them would let the system dismiss
anything it failed to parse.

Every category also carries ``excludedWhenAlso``: signals that disqualify the classification.
A swollen painful leg after birth is not a musculoskeletal complaint for our purposes, and a
fall in pregnancy with bleeding is emphatically not a sports injury.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path
from typing import Mapping

from app.context.target_entity_resolver import _contains_any, _fold
from app.rules.condition import Presence, signal_presence

TAXONOMY_PATH = Path(__file__).resolve().parents[2] / "data" / "oos_complaint_taxonomy_v1.json"


@lru_cache(maxsize=1)
def load_taxonomy() -> Mapping[str, object]:
    return json.loads(TAXONOMY_PATH.read_text(encoding="utf-8"))


@dataclass(frozen=True)
class ComplaintClassification:
    #: ``None`` when nothing matched — which means "unknown", never "out of scope".
    category_id: str | None
    evidence: tuple[str, ...] = ()
    #: Categories that matched textually but were disqualified by a present signal.
    disqualified: tuple[str, ...] = ()

    @property
    def is_confirmed_non_reproductive(self) -> bool:
        return self.category_id is not None


def classify_complaint(
    message: str | None,
    signals: Mapping[str, object] | None = None,
) -> ComplaintClassification:
    """Classify a complaint as belonging to a known non-reproductive domain, or not at all."""

    folded = _fold(message)
    if not folded:
        return ComplaintClassification(None)

    signals = signals or {}
    disqualified: list[str] = []

    for category in load_taxonomy()["categories"]:
        hits = _contains_any(folded, category["phrases"])
        if not hits:
            continue

        blocking = [
            code for code in category.get("excludedWhenAlso", ())
            if signal_presence(signals, code) is Presence.PRESENT
        ]
        if blocking:
            # Matched the words, but a signal present in this session makes the
            # classification unsafe — e.g. bleeding alongside a reported fall.
            disqualified.append(category["categoryId"])
            continue

        return ComplaintClassification(category["categoryId"], tuple(hits), tuple(disqualified))

    return ComplaintClassification(None, (), tuple(disqualified))


def may_return_out_of_scope(
    *,
    classification: ComplaintClassification,
    safety_screen_complete: bool,
    has_reproductive_evidence: bool,
) -> bool:
    """All three policy conditions must hold before OUT_OF_SCOPE may be returned.

    The safety-screen requirement is why "đau cổ tay" is not instantly out of scope: we must
    first know the person is not also short of breath.
    """

    policy = load_taxonomy()["policy"]
    if not classification.is_confirmed_non_reproductive:
        return False
    if policy.get("requiresCompleteGlobalSafetyScreen", True) and not safety_screen_complete:
        return False
    if policy.get("requiresNoReproductiveEvidence", True) and has_reproductive_evidence:
        return False
    return True

File: 05_Development/CareBridgeAPI/src/main/java/com/carebridge/backend/triage/rules/ComplaintTaxonomy.java
package com.carebridge.backend.triage.rules;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.core.io.ClassPathResource;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Deterministic out-of-scope complaint classification.
 *
 * <p>OUT_OF_SCOPE requires <em>positive evidence</em> that a complaint belongs to another
 * clinical domain. An unrecognised complaint is NEEDS_MORE_INFO — "we did not understand you"
 * is not the same statement as "this is not our field", and conflating them would let the
 * system dismiss anything it failed to parse.
 *
 * <p>Behavioural parity with {@code app/context/complaint_taxonomy.py}.
 */
@Component
public class ComplaintTaxonomy {

    public static final String TAXONOMY_RESOURCE = "triage/oos_complaint_taxonomy_v1.json";

    private final JsonNode taxonomy;

    public ComplaintTaxonomy() {
        this(TAXONOMY_RESOURCE);
    }

    ComplaintTaxonomy(String resourcePath) {
        try (InputStream stream = new ClassPathResource(resourcePath).getInputStream()) {
            this.taxonomy = new ObjectMapper().readTree(stream);
        } catch (IOException exception) {
            throw new IllegalStateException("cannot read complaint taxonomy " + resourcePath, exception);
        }
    }

    /** {@code categoryId} is null when nothing matched — which means unknown, not out of scope. */
    public record ComplaintClassification(
            String categoryId, List<String> evidence, List<String> disqualified) {

        public boolean isConfirmedNonReproductive() {
            return categoryId != null;
        }
    }

    public ComplaintClassification classify(String message, Map<String, Object> signals) {
        String folded = TargetEntityResolver.fold(message);
        if (folded.isEmpty()) {
            return new ComplaintClassification(null, List.of(), List.of());
        }

        List<String> disqualified = new ArrayList<>();
        for (JsonNode category : taxonomy.get("categories")) {
            List<String> hits = TargetEntityResolver.containsAny(folded, category.get("phrases"));
            if (hits.isEmpty()) {
                continue;
            }

            boolean blocked = false;
            for (JsonNode code : category.path("excludedWhenAlso")) {
                if (signals != null
                        && Presence.parse(signals.get(code.asText())) == Presence.PRESENT) {
                    blocked = true;
                    break;
                }
            }
            if (blocked) {
                // Matched the words, but a signal present in this session makes the
                // classification unsafe — e.g. bleeding alongside a reported fall.
                disqualified.add(category.get("categoryId").asText());
                continue;
            }
            return new ComplaintClassification(category.get("categoryId").asText(),
                    List.copyOf(hits), List.copyOf(disqualified));
        }
        return new ComplaintClassification(null, List.of(), List.copyOf(disqualified));
    }

    /**
     * All three policy conditions must hold before OUT_OF_SCOPE may be returned. The
     * safety-screen requirement is why "đau cổ tay" is not instantly out of scope: we must
     * first know the person is not also short of breath.
     */
    public boolean mayReturnOutOfScope(
            ComplaintClassification classification,
            boolean safetyScreenComplete,
            boolean hasReproductiveEvidence) {

        if (!classification.isConfirmedNonReproductive()) {
            return false;
        }
        JsonNode policy = taxonomy.path("policy");
        if (policy.path("requiresCompleteGlobalSafetyScreen").asBoolean(true)
                && !safetyScreenComplete) {
            return false;
        }
        return !(policy.path("requiresNoReproductiveEvidence").asBoolean(true)
                && hasReproductiveEvidence);
    }
}
